emu/src/cpu/instruction.rs

```diff
@@ -67,11 +187,77 @@ impl From<u32> for ArmModeInstruction {
         } else if op_code.get_bits(25..=27) == 0b100 {
             BlockDataTransfer
         } else if op_code.get_bits(25..=27) == 0b101 {
-            Branch
+            let is_link = op_code.get_bit(24);
+            let offset = op_code.get_bits(0..=23) << 2;
+            Branch(condition, is_link, offset)
         } else if op_code.get_bits(26..=27) == 0b01 {
-            SingleDataTransfer
+            let op_kind: OperandKind = op_code.get_bit(25).into();
+            let indexing: Indexing = op_code.get_bit(24).into(); // FIXME: should we use this?
+            let offsetting: Offsetting = op_code.get_bit(23).into();
+            let byte_or_word: ReadWriteKind = op_code.into(); // TODO: is this the same for all instruction?
+            let load_store: SingleDataTransferKind = op_code.into(); // TODO: is this the same bit for all instruction?
+            let write_back = op_code.get_bit(21);
+            let rn = op_code.get_bits(16..=19);
+            let rd = op_code.get_bits(12..=15);
+
+            let offset = match op_kind {
+                OperandKind::Immediate => todo!(),
+                OperandKind::Register => op_code.get_bits(0..=11),
+            };
```

Comment:
I think that we have to move this logic in the execute part because we can't read the register here in the decode.

Also, the logic here is inverted, if it is immediate we read `0..=11` and we use it as offset

Reply:
How can we calculate offset in ASM?

Reply:
If the shift is offset is immediate specified the offset is written in the opcode (`0..=11`), otherwise it is an `Rm` shifted by something and the shift part is similar to the one in data processing but without the register-specified amount

Reply:
But we can't calculate the offset in the `decode` because in the case of a Register specified, we should read the register and we don't do in the in the decode phase